### DISCUSSION TITLE
Use berry matter library in external non tasmota project

### DISCUSSION THREAD

**Discussion by S-A-M-J** [arendst/Tasmota#19845] [Support request]
Hi, 

i am working on a project where i would like to use matter but also run complex custom code on an esp32 c3. Now in order to do so the only ways i found to implement such a thing are the esp-idf with esp-matter or the berry implementation for tasmota. Now the esp-idf is pretty bad to use in my opinion and there are basically no libraries available for most things. The tasmota implementation of matter however seems very slick as it would seem like it could work with arduino code, and like the perfect base for a matter based project. However, i haven't been able to identify how, and if, it could be used in a separate project basically sort of as a library. With the solidification process of the berry files into c files, it has become really murky to me just how the matter files are used within tasmota and how one could perhaps use defined functions etc. 

Perhaps someone can point me in the right direction or even tell me if it is possible at all since i've spent countless hours trying to figure it out but seem to be at a dead end. 

Now i know this is somewhat off topic as it isn't really that tasmota related but i thought asking won't hurt.

Thx in advance

**Comment by s-hadinger** [comment]
Hi, unfortunately the Matter implementation is not designed to be a library for an Arduino project. It is tightly linked to how Tasmota works internally.

**Comment by sidbarg** [comment]
> Hi,
> 
> i am working on a project where i would like to use matter but also run complex custom code on an esp32 c3. Now in order to do so the only ways i found to implement such a thing are the esp-idf with esp-matter or the berry implementation for tasmota. Now the esp-idf is pretty bad to use in my opinion and there are basically no libraries available for most things. The tasmota implementation of matter however seems very slick as it would seem like it could work with arduino code, and like the perfect base for a matter based project. However, i haven't been able to identify how, and if, it could be used in a separate project basically sort of as a library. With the solidification process of the berry files into c files, it has become really murky to me just how the matter files are used within tasmota and how one could perhaps use defined functions etc.
> 
> Perhaps someone can point me in the right direction or even tell me if it is possible at all since i've spent countless hours trying to figure it out but seem to be at a dead end.
> 
> Now i know this is somewhat off topic as it isn't really that tasmota related but i thought asking won't hurt.
> 
> Thx in advance

@S-A-M-J have you meanwhile found any better way of achieving this ?

**Reply by S-A-M-J** [reply]
Well I went with using the esp matter sdk. It's really badly documented but once you understand the basics it's actually quite usable.
